File: frontend/src/types/mcp.ts
/**
 * MCP 服务器类型定义
 *
 * 契约真相源是后端 schemas/mcp.py（→ api.generated.ts）；
 * 底部 SameShape 锚点锁住手写类型与生成物逐字段一致（含可选性与 null），
 * 后端契约变更而前端没跟上时，本文件编译期报红。
 */

import type { components } from '@/types/api.generated'

// ============================================================================
// 连接状态
// ============================================================================

export type MCPConnectionStatus = 'unknown' | 'connected' | 'error'

/**
 * MCP 传输协议类型
 */
export type MCPTransport = 'sse' | 'streamable_http'

// ============================================================================
// 数据模型
// ============================================================================

/**
 * MCP 服务器接口
 */
export interface MCPServer {
  id: string
  name: string
  description: string | null
  sse_url: string
  transport: MCPTransport
  is_active: boolean
  icon: string | null
  connection_status: MCPConnectionStatus
  created_at: string
  updated_at: string
}

/**
 * 创建 MCP 服务器的 DTO
 */
export interface MCPServerCreate {
  name: string
  description?: string | null
  sse_url: string
  transport: MCPTransport
  icon?: string | null
}

/**
 * 更新 MCP 服务器的 DTO
 */
export interface MCPServerUpdate {
  name?: string | null
  description?: string | null
  sse_url?: string | null
  transport?: MCPTransport | null
  is_active?: boolean | null
  icon?: string | null
}

// ============================================================================
// REST 契约锚点（范式沿 types/events.ts 的 SameShape / services/stats.ts）
// ============================================================================

/** 双向相等：手写类型与后端生成类型**逐字段一致**（含可选性与 null）。 */
type SameShape<A, B> = [A] extends [B] ? ([B] extends [A] ? true : false) : false
type Assert<T extends true> = T

type _MCPServer = Assert<SameShape<MCPServer, components['schemas']['MCPServerResponse']>>
type _MCPServerCreate = Assert<
  SameShape<MCPServerCreate, components['schemas']['MCPServerCreate']>
>
type _MCPServerUpdate = Assert<
  SameShape<MCPServerUpdate, components['schemas']['MCPServerUpdate']>
>

/** 上面这组断言只做编译期校验，导出以免被 noUnusedLocals 误报 */
export type McpConformanceAnchors = [_MCPServer, _MCPServerCreate, _MCPServerUpdate]

File: backend/schemas/mcp.py
"""MCP 服务器 DTO。

从 models/mcp.py 迁入（架构约定：ORM 表模型归 models/，DTO 归 schemas/）。
transport / connection_status 的取值空间在此用 Literal 声明——它们同时是
OpenAPI 契约的一部分（前端 mcp.ts 的同名联合类型与这里逐字对齐，
SameShape 锚点锁定）。
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict
from pydantic import Field as PydanticField

MCPTransport = Literal["sse", "streamable_http"]
MCPConnectionStatus = Literal["unknown", "connected", "error"]


class MCPServerCreate(BaseModel):
    """创建 MCP 服务器的 DTO

    添加新服务器时，后端会执行连接测试，
    只有连接成功才会入库。
    """

    name: str = PydanticField(..., min_length=1, max_length=100, description="显示名称")
    description: str | None = PydanticField(default=None, max_length=500, description="功能描述")
    sse_url: str = PydanticField(..., description="MCP 服务器连接地址")
    # 不带 | None：缺省即 sse（路由侧 or "sse" 同义），显式 null 会被 422 拒绝，
    # 契约保持「可选、非空」而不是「必填、可空」
    transport: MCPTransport = PydanticField(
        default="sse", description="传输协议：sse 或 streamable_http"
    )
    icon: str | None = PydanticField(default=None, description="图标")


class MCPServerUpdate(BaseModel):
    """更新 MCP 服务器的 DTO

    支持部分更新，所有字段均为可选。
    """

    name: str | None = PydanticField(default=None, min_length=1, max_length=100)
    description: str | None = PydanticField(default=None, max_length=500)
    sse_url: str | None = None
    transport: MCPTransport | None = None
    is_active: bool | None = None
    icon: str | None = None


class MCPServerResponse(BaseModel):
    """MCP 服务器响应 DTO

    包含 connection_status 字段供前端展示状态灯。
    """

    id: str
    name: str
    description: str | None
    sse_url: str
    transport: MCPTransport
    is_active: bool
    icon: str | None
    connection_status: MCPConnectionStatus
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
